lib/fine-types/src/Language/FineTypes/Export/OpenAPI/Value/ToJSON.hs

```diff
@@ -47,28 +49,41 @@ import qualified Language.FineTypes.Typ as Typ
 -- We need the 'Typ' of the 'Value' in order to add field names.
 --
 -- Note: We assume that the 'Value' has the given 'Typ'.
+reportTypeValue :: (Show t, Show v) => t -> v -> String
+reportTypeValue typ value =
+    TL.unpack $ pShow (("type:" :: Text, typ), ("value:" :: Text, value))
+
+typeCheckAssertion :: Typ -> Value -> t -> t
+typeCheckAssertion typ value t =
+    if hasTyp value typ
+        then t
+        else error $ "typeCheckAssertion: " <> reportTypeValue typ value
+
+impossibleCase :: (Show t, Show v) => t -> v -> String -> a
+impossibleCase typ value prefix =
+    error $ prefix <> ": " <> reportTypeValue typ value
+
 jsonFromValue :: Typ -> Value -> JS.Value
-jsonFromValue = go
+jsonFromValue t0 v0 = typeCheckAssertion t0 v0 $ go t0 v0
   where
     go :: Typ -> Value -> JS.Value
     go (Typ.Two Typ.Product2 ta tb) (Product [a, b]) =
         JS.object ["0" .= go ta a, "1" .= go tb b]
-    go (Typ.ProductN fields) x@(Product _) =
-        jsonFromProduct fields
-            $ flattenBinaryProduct (length fields) x
-    -- TODO: Don't flatten automatically when Embedding get better.
-
+    go (Typ.ProductN fields) (Product ps) = jsonFromProduct fields ps
     go (Typ.Two Typ.Sum2 ta _) (Sum 0 a) =
         JS.object ["0" .= go ta a]
     go (Typ.Two Typ.Sum2 _ tb) (Sum 1 b) =
         JS.object ["1" .= go tb b]
     go (Typ.SumN constructors) (Sum ix a) =
         jsonFromSum constructors ix a
-    go Typ.Abstract _ = error "jsonFromValue: Typ may not be abstract."
+    go Typ.Abstract v =
+        impossibleCase Typ.Abstract v "Only concrete types"
     go (Typ.Var _) (Zero v) = go0 v
     go (Typ.One op t) (One v) = go1 op t v
-    go Typ.Two{} (Two v) = go2 v
-    go _ _ = error "jsonFromValue: Typ error"
+    go t@Typ.Two{} v@(Two _) =
+        impossibleCase t v "FiniteMapV is not supported by JSON"
+    go (Typ.Zero _t) (Zero v) = go0 v
+    go t v = impossibleCase t v "Impossible case"
```

Comment:
```suggestion
    go t v = impossibleCase t v "Value does not type check"
```

This should be a good way of reporting the type checking failure without incurring the additional run-time cost of `hasTyp`. 🤔

Reply:
We have no proof that this is the same as hasTyp, but I agree there is an additional cost